File: src/types/manage.ts
import type { DataSource } from './track'

/** Filter parameters for querying track metadata.
 *  All fields optional — empty/undefined = show all. */
export interface TrackMetaFilter {
  source?: DataSource
  searchText?: string
  batchId?: number
  airline?: string
  aircraftType?: string
  minPoints?: number
  maxPoints?: number
  minTimeMs?: number
  maxTimeMs?: number
  sortBy?: string
  sortDesc?: boolean
}

/** Lightweight track metadata row (matches Rust TrackMetaInfo) */
export interface TrackMetaInfo {
  icao_address: string
  batch_id: number
  source: string
  flight_number: string
  icao_flight_number: string
  registration: string
  aircraft_type: string
  airline: string
  origin: string
  destination: string
  min_timestamp: string
  max_timestamp: string
  point_count: number
  batch_file_name: string
  batch_imported_at: string
}

/** Aggregated statistics */
export interface TrackStats {
  total_tracks: number
  total_batches: number
  by_source: Record<string, number>
  unique_icao: number
  unique_airlines: string[]
  unique_aircraft_types: string[]
  time_min_ms: number | null
  time_max_ms: number | null
}

/** Distinct values for filter dropdowns */
export interface DistinctOptions {
  airlines: string[]
  aircraft_types: string[]
  batch_names: [number, string][]
}

/** Paginated response */
export interface TrackMetadataResponse {
  rows: TrackMetaInfo[]
  total_count: number
}

/** Sort state */
export interface SortConfig {
  column: string
  desc: boolean
}

/** Default filter — all empty = show everything */
export function defaultFilter(): TrackMetaFilter {
  return { sortBy: 'batch_imported_at', sortDesc: true }
}

export const DEFAULT_SORT: SortConfig = { column: 'batch_imported_at', desc: true }

/** Source dropdown options */
export const SOURCE_OPTIONS: { value: string; label: string }[] = [
  { value: 'adsb', label: 'ADS-B' },
  { value: 'radar', label: '雷达' },
  { value: 'radar_raw', label: '雷达原始' },
]

File: src-tauri/src/manage.rs
use serde::{Deserialize, Serialize};

/// Map frontend DataSource values to DB source column values.
pub fn source_to_db(frontend_source: &str) -> &str {
    match frontend_source {
        "adsb" => "ADS-B",
        "radar" => "Radar",
        "radar_raw" => "RadarRaw",
        "simulation" => "Simulation",
        other => other,
    }
}

/// Map DB source column values to frontend DataSource values.
pub fn source_from_db(db_source: &str) -> &str {
    match db_source {
        "ADS-B" => "adsb",
        "Radar" => "radar",
        "RadarRaw" => "radar_raw",
        "Simulation" => "simulation",
        other => other,
    }
}

/// Filter parameters for querying track metadata.
/// `source` is optional — when None, all sources are included.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrackMetaFilter {
    #[serde(default)]
    pub source: Option<String>,
    #[serde(default)]
    pub search_text: Option<String>,
    #[serde(default)]
    pub batch_id: Option<i64>,
    #[serde(default)]
    pub airline: Option<String>,
    #[serde(default)]
    pub aircraft_type: Option<String>,
    #[serde(default)]
    pub min_points: Option<i64>,
    #[serde(default)]
    pub max_points: Option<i64>,
    #[serde(default)]
    pub min_time_ms: Option<i64>,
    #[serde(default)]
    pub max_time_ms: Option<i64>,
    #[serde(default = "default_sort_by")]
    pub sort_by: String,
    #[serde(default)]
    pub sort_desc: bool,
}

fn default_sort_by() -> String {
    "batch_imported_at".to_string()
}

/// Lightweight metadata row for the management table (no position data).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrackMetaInfo {
    pub icao_address: String,
    pub batch_id: i64,
    pub source: String,
    pub flight_number: String,
    pub icao_flight_number: String,
    pub registration: String,
    pub aircraft_type: String,
    pub airline: String,
    pub origin: String,
    pub destination: String,
    pub min_timestamp: String,
    pub max_timestamp: String,
    pub point_count: i64,
    pub batch_file_name: String,
    pub batch_imported_at: String,
}

/// Aggregated statistics across all tracks.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrackStats {
    pub total_tracks: i64,
    pub total_batches: i64,
    pub by_source: std::collections::HashMap<String, i64>,
    pub unique_icao: i64,
    pub unique_airlines: Vec<String>,
    pub unique_aircraft_types: Vec<String>,
    pub time_min_ms: Option<i64>,
    pub time_max_ms: Option<i64>,
}

/// Distinct values for filter dropdowns.
/// When `source` is None, returns values across all sources.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DistinctOptions {
    pub airlines: Vec<String>,
    pub aircraft_types: Vec<String>,
    /// (batch_id, file_name) pairs
    pub batch_names: Vec<(i64, String)>,
}

/// Wrapper for the paginated query response sent to the frontend.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrackMetadataResponse {
    pub rows: Vec<TrackMetaInfo>,
    pub total_count: i64,
}
